clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala

```diff
@@ -182,29 +180,35 @@ object UncommittedGarbageCollector {
           .repartition(dataDF.col("address"))
           .except(committedDF)
           .except(uncommittedDF)
+          .cache()
+
+        committedDF.unpersist()
+        uncommittedDF.unpersist()
+      }
+
+      // write addresses to delete
+      if (shouldMark) {
+        val reportDst = formatRunPath(storageNamespace, runID)
+        println(s"Report for mark_id=$runID path=$reportDst")
+
+        addressesToDelete.write.parquet(s"$reportDst/deleted")
+        addressesToDelete.write.text(s"$reportDst/deleted.text")
       }
-      var jobID = runID
+      // delete marked addresses
       if (shouldSweep) {
-        if (shouldMark) { // get the expired addresses from the mark id run
-          markedAddresses = addressesToDelete
-        } else {
-          jobID = markID
-          markedAddresses = readMarkedAddresses(storageNamespace, markID)
-        }
-        println("deleting marked addresses from job: " + jobID)
+        val jobID = if (shouldMark) runID else markID
+        println("deleting marked addresses from job ID: " + jobID)
+        val markedAddresses = readMarkedAddresses(storageNamespace, jobID)
```

Comment:
If I understand correctly, now we are writing and reading the marked addresses every time we run the GC. What is the impact performance-wise?

Reply:
Still testing the two options:
- use the cached results
- use the dataframe we just saved

Will update the PR with the preferred way.

Reply:
@N-o-Z updated the code it reuse the cached dataset